File: user_interface/lib/domain/entities/annotation.dart
// lib/domain/entities/annotation.dart
//
// ReviewCard/DeepNoteのmetadata.annotations配列の1要素。
// ハイライト(highlight)やノート(notes)など、ユーザーが本文の一部に付ける
// 注釈を表す。位置はblockIdx(Review Cardsのみ、DeepNotesはnull)+start/end_idx
// で特定し、annotatedWordsは表示コンテンツが変わった際の検証・再アンカリング
// 用のアンカーテキストとして持つ。

class Annotation {
  const Annotation({
    required this.id,
    this.blockIdx,
    required this.startIdx,
    required this.endIdx,
    required this.annotationType,
    required this.annotatedWords,
    required this.contents,
  });

  final String id;
  // Review Cardsのcard_content内のブロックindex。DeepNotesはnull。
  final int? blockIdx;
  final int startIdx;
  final int endIdx;
  // "highlight" | "notes"
  final String annotationType;
  // 保存時点で選択されていたテキスト(検証・再アンカリング用のアンカー)
  final String annotatedWords;
  // highlight: {"type": "marker"|"wave", "color": "..."}
  // notes: ノート本文の文字列
  final dynamic contents;

  factory Annotation.fromMap(Map<String, dynamic> map) {
    return Annotation(
      id: map['id'] as String,
      blockIdx: (map['block_idx'] as num?)?.toInt(),
      startIdx: (map['start_idx'] as num?)?.toInt() ?? 0,
      endIdx: (map['end_idx'] as num?)?.toInt() ?? 0,
      annotationType: map['annotation_type'] as String? ?? 'highlight',
      annotatedWords: map['annotated_words'] as String? ?? '',
      contents: map['contents'],
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      if (blockIdx != null) 'block_idx': blockIdx,
      'start_idx': startIdx,
      'end_idx': endIdx,
      'annotation_type': annotationType,
      'annotated_words': annotatedWords,
      'contents': contents,
    };
  }
}

File: webapp/src/types/annotation.ts
/**
 * review_cards / deep_notes の metadata.annotations 配列の1要素。
 * user_interface/lib/domain/entities/annotation.dart と同一スキーマ。
 *
 * start_idx / end_idx は「Markdown記法とSID引用を取り除いたフラット化テキスト」
 * 上のオフセット。詳しくは lib/markdownAnnotations.ts を参照。
 */
export type AnnotationType = 'highlight' | 'notes';

/** highlight用の contents。marker=背景塗り, line=下線, wave=波線。 */
export interface HighlightContents {
  type: 'marker' | 'line' | 'wave';
  color: string;
}

/** notes の contents はノート本文の文字列そのもの。 */
export type AnnotationContents = HighlightContents | string | null;

export interface Annotation {
  id: string;
  /** ReviewCardのcard_content内のブロックindex。DeepNotesはnull/undefined。 */
  block_idx?: number | null;
  start_idx: number;
  end_idx: number;
  annotation_type: AnnotationType;
  /** 保存時点で選択されていたテキスト(検証・再アンカリング用)。 */
  annotated_words: string;
  contents: AnnotationContents;
}

export const HIGHLIGHT_PRESET_COLORS = [
  '#FFB300', // Star Gold
  '#EF5350', // Red
  '#F06292', // Pink
  '#BA68C8', // Purple
  '#7986CB', // Indigo
  '#64B5F6', // Blue
  '#4DB6AC', // Teal
  '#81C784', // Green
  '#FF9800', // Orange
] as const;

export const NOTE_MARKER_COLOR = '#C5A059';

export function isHighlightContents(contents: AnnotationContents): contents is HighlightContents {
  return typeof contents === 'object' && contents !== null && 'type' in contents;
}

export function noteText(annotation: Annotation): string {
  return typeof annotation.contents === 'string' ? annotation.contents : '';
}

export function createAnnotationId(): string {
  return crypto.randomUUID();
}
